Variational Inference for Dirichlet Process Gaussian Mixture

mamikonyana opened the pull request:
This also has some helper classes for diagnostics, I could separate into a differnet pull requets if that will make it easier to go over the diff.

The main thing that this diff accomplishes is that now we can infer an infinite mixture of Gaussians when doing density estimation. It's slow, but it's a good start.

mamikonyana commented:
@pbailis can we change the coverage check to be slightly more tolerable?

viveksjain commented on src/main/java/macrobase/diagnostics/ScoreDumper.java:
```diff
@@ -0,0 +1,41 @@
+package macrobase.diagnostics;
```
This file caught my eye - is there a particular reason this couldn't be integrated with src/test/java/macrobase/diagnostic/tasks/ScoreDumpDiagnostic.java?

mamikonyana replied:
yes actually. I had some code that would run diagnostics not from tests. I still want to, because sometimes all i need is run macrobase on a given dataset and dump the results in a file to be able to plot. Currently my solution is to write a test for that case, which does not give me freedom to automatically iterate over configuration parameters or datasets. So I think there should be a diagnostics pipeline or switch in the main package, I can't decide which one would be nicer and more maintainable.

mamikonyana replied:
Also thanks for looking through the code :)

pbailis replied:
@edgan8 talked about having a scoredump operator that you could stick into the pipeline. what do you think about that?

viveksjain replied:
Sorry I didn't get to this earlier, but ScoreDumpDiagnostic can be run as a pipeline, right? See `bin/diagnostic.sh`.

pbailis commented:
> can we change the coverage check to be slightly more tolerable?

What do you mean?

pbailis commented on src/main/java/macrobase/analysis/stats/mixture/VariationalDPMG.java:
```diff
@@ -0,0 +1,288 @@
+            log.info("log likelihood after iteration {} is {}", iteration, logLikelihood);
+
+            if (iteration >= maxIterationsToConverge) {
+                log.info("Breaking because have already run {} iterations", iteration);
```
Shouldn't these be log.debug or log.trace?

pbailis commented on src/main/java/macrobase/analysis/stats/mixture/VariationalDPMG.java:
```diff
@@ -0,0 +1,288 @@
+                logLikelihood += Math.log(score(data.get(n)));
+            }
+
+            log.info("log likelihood after iteration {} is {}", iteration, logLikelihood);
```
Here too.

pbailis commented on src/main/java/macrobase/analysis/stats/mixture/VariationalDPMG.java:
```diff
@@ -0,0 +1,288 @@
+
+            double improvement = (logLikelihood - oldLogLikelihood) / ( -logLikelihood);
+            if (improvement >= 0 && improvement < this.progressCutoff) {
+                log.debug("Breaking because improvement was {} percent", improvement * 100);
```
Here too.

pbailis commented on src/main/java/macrobase/analysis/stats/mixture/VariationalDPMG.java:
```diff
@@ -0,0 +1,288 @@
+        }
+    }
+
+    private void instanciateVariationalParameters(int numPoints, int maxNumClusters) {
```
typo: instantiate

pbailis commented on src/main/java/macrobase/analysis/transform/BatchScoreFeatureTransform.java:
```diff
@@ -28,8 +32,28 @@ public BatchScoreFeatureTransform(MacroBaseConf conf, Iterator<Datum> input) thr
         for(Datum d : data) {
+            double score = batchTrainScore.score(d);
             results.add(new Datum(d, batchTrainScore.score(d)));
+            if (dumpScores) {
```
@edgan8 is this what you were thinking of doing?

pbailis commented on src/main/java/macrobase/analysis/transform/BatchScoreFeatureTransform.java:
```diff
@@ -28,8 +32,28 @@ public BatchScoreFeatureTransform(MacroBaseConf conf, Iterator<Datum> input) thr
+                metricsAndDensities.add(new MetricsAndDensity(d.getMetrics(), score));
+            }
+        }
+        if (dumpScores) {
```
Is it possible to do this inside a different utility class? @edgan8 had some ideas.

pbailis commented:
Coverage decrease is not phenomenal... :)

pbailis commented:
LGTM, with a few concerns to address beforehand:

 - Not sure why so many of the log entries need to be at `info`.
 - Would be good to discuss how to dump the scores. @edgan8 had some thoughts.
 - If Travis is failing spuriously, can we use smaller datasets for the tests?

mamikonyana commented:
>> can we change the coverage check to be slightly more tolerable?

>What do you mean?

Make coverage check not fail if coverage decreased by less than say 0.1% :)

mamikonyana commented:
> - If Travis is failing spuriously, can we use smaller datasets for the tests?

Travis does not fail any longer after I introduced a limit on number of iterations tests run

>  - Not sure why so many of the log entries need to be at info.

Changed info's into debugs

>  - Would be good to discuss how to dump the scores. @edgan8 had some thoughts.

I reverted `src/main/java/macrobase/analysis/transform/BatchScoreFeatureTransform.java`, I though I had done that before too. I'll talk to @edgan8 about his ideas.

pbailis commented:
> Make coverage check not fail if coverage decreased by less than say 0.1% :)

1000 * 0.1% = 100%!
